# Machine Learning HW1

## PM2.5 Prediction

In [1]:
import numpy as np
import pandas as pd

In [2]:
data = pd.read_csv('ml2020spring-hw1/train.csv', encoding = 'big5')
data.head(18)

日期  測站          測項     0     1     2     3     4     5     6  ...  \
0   2014/1/1  豐原    AMB_TEMP    14    14    14    13    12    12    12  ...   
1   2014/1/1  豐原         CH4   1.8   1.8   1.8   1.8   1.8   1.8   1.8  ...   
2   2014/1/1  豐原          CO  0.51  0.41  0.39  0.37  0.35   0.3  0.37  ...   
3   2014/1/1  豐原        NMHC   0.2  0.15  0.13  0.12  0.11  0.06   0.1  ...   
4   2014/1/1  豐原          NO   0.9   0.6   0.5   1.7   1.8   1.5   1.9  ...   
5   2014/1/1  豐原         NO2    16   9.2   8.2   6.9   6.8   3.8   6.9  ...   
6   2014/1/1  豐原         NOx    17   9.8   8.7   8.6   8.5   5.3   8.8  ...   
7   2014/1/1  豐原          O3    16    30    27    23    24    28    24  ...   
8   2014/1/1  豐原        PM10    56    50    48    35    25    12     4  ...   
9   2014/1/1  豐原       PM2.5    26    39    36    35    31    28    25  ...   
10  2014/1/1  豐原    RAINFALL    NR    NR    NR    NR    NR    NR    NR  ...   
11  2014/1/1  豐原          RH    77    68    67    74    72    73    74  ...   
12  2014/1/1  豐原         SO2   1.8     2   1.7   1.6   1.9   1.4   1.5  ...   
13  2014/1/1  豐原         THC     2     2     2   1.9   1.9   1.8   1.9  ...   
14  2014/1/1  豐原       WD_HR    37    80    57    76   110   106   101  ...   
15  2014/1/1  豐原  WIND_DIREC    35    79   2.4    55    94   116   106  ...   
16  2014/1/1  豐原  WIND_SPEED   1.4   1.8     1   0.6   1.7   2.5   2.5  ...   
17  2014/1/1  豐原       WS_HR   0.5   0.9   0.6   0.3   0.6   1.9     2  ...   

      14    15    16    17    18    19    20    21    22    23  
0     22    22    21    19    17    16    15    15    15    15  
1    1.8   1.8   1.8   1.8   1.8   1.8   1.8   1.8   1.8   1.8  
2   0.37  0.37  0.47  0.69  0.56  0.45  0.38  0.35  0.36  0.32  
3    0.1  0.13  0.14  0.23  0.18  0.12   0.1  0.09   0.1  0.08  
4    2.5   2.2   2.5   2.3   2.1   1.9   1.5   1.6   1.8   1.5  
5     11    11    22    28    19    12   8.1     7   6.9     6  
6     14    13    25    30    21    13   9.7   8.6   8.7   7.5  
7     65    64    51    34    33    34    37    38    38    36  
8     52    51    66    85    85    63    46    36    42    42  
9     36    45    42    49    45    44    41    30    24    13  
10    NR    NR    NR    NR    NR    NR    NR    NR    NR    NR  
11    47    49    56    67    72    69    70    70    70    69  
12   3.9   4.4   9.9   5.1   3.4   2.3     2   1.9   1.9   1.9  
13   1.9   1.9   1.9   2.1     2   1.9   1.9   1.9   1.9   1.9  
14   307   304   307   124   118   121   113   112   106   110  
15   313   305   291   124   119   118   114   108   102   111  
16   2.5   2.2   1.4   2.2   2.8     3   2.6   2.7   2.1   2.1  
17   2.1   2.1   1.9     1   2.5   2.5   2.8   2.6   2.4   2.3  

[18 rows x 27 columns]

In [3]:
data1 = data.iloc[:,3:]
data1[data1 == 'NR'] = 0
data1.shape

(4320, 24)

In [4]:
raw_data = data1.to_numpy()

### Extract Feature 1

将 $ 4320 * 18$ 的数据转为 $12 * 20 * 24 * 18 $ 的数据，12个月，每月20天，每天24小时，每小时18个特征。

必须分开12个月计算，每个月的时间不连续。

In [5]:
month_data = {}
for month in range(12):
    sample = np.empty([18, 480])
    for day in range(20):
        sample[:,day * 24 : (day+1) * 24] = raw_data[18 * (month * 20 + day) : 18 * (month * 20 + day + 1),:]
    month_data[month] = sample

In [6]:
# 每个月数据均为 18 * 480
month_data[11]

array([[ 23.  ,  23.  ,  23.  , ...,  13.  ,  13.  ,  13.  ],
       [  1.6 ,   1.7 ,   1.7 , ...,   1.8 ,   1.8 ,   1.8 ],
       [  0.22,   0.2 ,   0.18, ...,   0.51,   0.57,   0.56],
       ...,
       [ 93.  ,  50.  ,  99.  , ..., 118.  , 100.  , 105.  ],
       [  1.8 ,   2.1 ,   3.2 , ...,   1.5 ,   2.  ,   2.  ],
       [  1.3 ,   0.9 ,   1.  , ...,   1.6 ,   1.8 ,   2.  ]])

### Extract Feature 2
每个月连续时间480小时，每9个小时产生一组data，每月共471组连续数据。

In [7]:
x_data = np.empty([471 * 12, 18 * 9], dtype = float)
y_data = np.empty([471 * 12, 1], dtype = float)
for month in range(12):
    for day in range(20):
        for hour in range(24):
            if day == 19 and hour > 14:
                continue
            x_data[month * 471 + day * 24 + hour,:] = month_data[month][:,day * 24 + hour : day * 24 + hour + 9].reshape(1, -1)
            y_data[month * 471 + day * 24 + hour,:] = month_data[month][9,day * 24 + hour + 9]
print(x_data.shape)
print(y_data.shape)

(5652, 162)
(5652, 1)


### Normalization

In [8]:
x_mean = np.mean(x_data, axis = 0)
x_std = np.std(x_data, axis = 0)
x_mean.shape,x_std.shape

((162,), (162,))

In [9]:
for i in range(len(x_data)):
    for j in range(len(x_data[i])):
        if x_std[j]:
            x_data[i][j] = (x_data[i][j] - x_mean[j]) / x_std[j]
x_data

array([[-1.35825331, -1.35883937, -1.359222  , ...,  0.26650729,
         0.2656797 , -1.14082131],
       [-1.35825331, -1.35883937, -1.51819928, ...,  0.26650729,
        -1.13963133, -1.32832904],
       [-1.35825331, -1.51789368, -1.67717656, ..., -1.13923451,
        -1.32700613, -0.85955971],
       ...,
       [-0.88092053, -0.72262212, -0.56433559, ..., -0.57693779,
        -0.29644471, -0.39079039],
       [-0.7218096 , -0.56356781, -0.72331287, ..., -0.29578943,
        -0.39013211, -0.1095288 ],
       [-0.56269867, -0.72262212, -0.88229015, ..., -0.38950555,
        -0.10906991,  0.07797893]])

### Shuffle the set

In [59]:
import random
index = [i for i in range(len(x_data))]
random.shuffle(index)
x_data = x_data[index]

### Split Validation Set

In [60]:
import math
x_train, y_train = x_data[:math.floor(len(x_data) * 0.8),:], y_data[:math.floor(len(x_data) * 0.8),:]
x_val, y_val = x_data[math.floor(len(x_data) * 0.8):,:], y_data[math.floor(len(x_data) * 0.8):,:]

### Linear Regression

### Adagrad

In [63]:
iteration = 3000
learning_rate = 100
eps = 1e-7
rows = int(x_train.shape[0])
params = int(x_train.shape[1])

w = np.matrix(np.zeros([params+1, 1]))
x = np.matrix(np.concatenate((np.ones([rows, 1]),x_train), axis = 1).astype(float))
y = np.matrix(y_train)

adgrad = np.matrix(np.zeros([params+1, 1]))
for item in range(iteration):
    z = x * w - y
    loss = np.sum(np.power((x * w - y), 2)) / (rows * 2)
    gradient = x.T * (x * w - y) / rows
    adgrad += np.power(gradient, 2)
    w = w - np.multiply(learning_rate / (np.sqrt(adgrad) + eps) , gradient)
    if (item+1) % 100 == 0:
        print(f'No.{item+1} loss is {loss}')

No.100 loss is 317789.39911053376
No.200 loss is 113778.16286938918
No.300 loss is 58422.67407566133
No.400 loss is 34747.065315552616
No.500 loss is 22386.362672656123
No.600 loss is 15138.24245595807
No.700 loss is 10573.734165545917
No.800 loss is 7564.170371096305
No.900 loss is 5516.2309618577
No.1000 loss is 4089.8506443231886
No.1100 loss is 3078.0947616085286
No.1200 loss is 2349.595916282807
No.1300 loss is 1818.313374246674
No.1400 loss is 1426.514842934186
No.1500 loss is 1134.6981629163583
No.1600 loss is 915.3890419206072
No.1700 loss is 749.2084235921003
No.1800 loss is 622.3182598547368
No.1900 loss is 524.729002703214
No.2000 loss is 449.1587054981753
No.2100 loss is 390.25233264655583
No.2200 loss is 344.0403786404491
No.2300 loss is 307.5589127111676
No.2400 loss is 278.58000836097665
No.2500 loss is 255.4185945381079
No.2600 loss is 236.79281534460966
No.2700 loss is 221.72224426895477
No.2800 loss is 209.453133680352
No.2900 loss is 199.40314139919678
No.3000 loss i